onnx_caffe2/backend.py

```diff
@@ -226,40 +234,102 @@ def _create_constant(cls, node_def, env):
+    # operators inherit from "ConvPoolOpBase", which is an abstract class that
+    # defines all of the attributes (kernels, dilations, strides, etc) which one
+    # sees on these operators.  Unfortunately, Caffe2's documentation generator
+    # doesn't know how to handle cases like this, so for example, if you look at
+    # the docs for MaxPool at <https://caffe2.ai/docs/operators-catalogue.html#maxpool>
+    # you won't see any of the attributes.  You have to go source diving to
+    # find the information; in particular, you want to look at:
+    # https://github.com/caffe2/caffe2/blob/master/caffe2/operators/conv_pool_op_base.h
+    # This class handles *global* pooling as well.
+    #
+    # Second, it's important to know what Caffe2 expects for padding, which can
+    # be somewhat difficult to understand from the code because Caffe2 handles
+    # both singular/pluralized spellings of padding, and there is also legacy
+    # padding business.  The short version of the story is that, for NON-legacy
+    # padding (which is what we want to output), padding is expected to be
+    # *twice* the size of kernels.  So if you have a 2D convolution, Caffe2
+    # will accept two values in 'kernels', but FOUR values in 'pads';
+    # furthermore, this is *mandatory.*
+    #
+    # Finally, ConvPoolOpBase is not the only class of it's kind; there is
+    # also ConvTransposeUnpoolBase, which backs ConvTranspose.  So don't
+    # be tricked by the fact that Conv and ConvTranspose have similar
+    # parameters; they exercise different codepaths and need to be handled
+    # differently.
+
     @classmethod
-    def _create_transpose(cls, node_def, env):
-        op_def = caffe2_pb2.OperatorDef()
-        op_def.output.extend([env[o] for o in node_def.output])
-        op_def.input.extend([env[i] for i in node_def.input])
-        op_def.type = 'ConvTranspose'
-        op_def.name = node_def.name
+    def _create_conv_pool_op_base(cls, node_def, env):
+        op_def = cls._common_op_translator(node_def, env)
+
+        # NB: Don't forget to rename the operator to the correct thing in
+        # _renamed_operators
+        if node_def.op_type.startswith('Global'): # test against unrenamed!
+            global_pooling_attr = op_def.arg.add()
+            global_pooling_attr.name = 'global_pooling'
+            global_pooling_attr.i = 1
+
+        # TODO: Postel was wrong
+        # https://tools.ietf.org/html/draft-thomson-postel-was-wrong-00
+        # Please apply more stringent checks on these in onnx repo.
+
+        d = dict(map(lambda arg: (arg.name, arg), op_def.arg))
+        if 'kernels' in d and 'pads' in d and len(d['kernels'].ints) == len(d['pads'].ints):
+            # Caffe2 requires pads to be twice the size of kernels.
+            # TODO: Hmm, this is going to mutate the original op_def.arg, isn't
+            # it...
+            new_pads = list(d['pads'].ints) * 2
```

Comment:
Should check whether all items have save values.

Reply:
I don't understand this comment. "save values"; did you mean "same"? But the point is that there weren't enough pads passed, so we're increasing it to fill; there's nothing to check.

Reply:
Yeah I meant "same" values, sorry for the typo.
The order of `pads` matters in caffe2, it should be top, left, bottom and right. Say the pads in the operator def has two ints, user can mean [top, left] or [left, top], and if these two ints are different, extending the list by `*2` might cause surprising result. We need to either document the order or only extend it when the values are the same.

Reply:
So, at least in PyTorch, there is a well defined meaning for two padding values: it's `top, left`. I think ONNX should adopt this convention; it matches with the meaning of dimensions in the Tensor.